### PR TITLE
Reduce memory usage by removing unnecessary regex

### PR DESCRIPTION
**Pull request by fukusuket**
## What Changed
- Reduced memory usage(about 200-300MB 
in the data below) by removing regex compilation process when the search can be replaced with the fast search below.
  - https://github.com/Yamato-Security/hayabusa/pull/884
  - https://github.com/Yamato-Security/hayabusa/pull/883

- Changed pipe startswith/endswith matching from regex to https://github.com/Yamato-Security/hayabusa/pull/890 match.
  - This change didn't speed up matching, but it is changed for matching process refactoring

### Environment
- OS: Windows 10 Home edition
- Hard: Memory 16GB , Core 8, SSD, laptop

### Benchmark1
I ran a benchmark using [this procedure(6.1GB evtx)](https://github.com/Yamato-Security/hayabusa/issues/778#issuecomment-1296504766) and the results were as follows.

|Version| Elapsed time | Memory(peak)  | Events with hits / Total events | Output file size(bytes)|
|-------| ------------ | --------|--------|--------| 
| [before](https://github.com/Yamato-Security/hayabusa/tree/imporve-speed-by-using-starts-ends-with)  | 00:10:44.026| 4.7 GiB | 1,593,715 / 4,817,181 | 575085389 |
| [This PR](https://github.com/Yamato-Security/hayabusa/tree/improve-speed-by-reduce-regex-match-when-pipe)   | 00:10:42.195| 4.4 GiB | 1,593,715 / 4,817,181 | 575085385 | 

#### Console output

[before](https://github.com/Yamato-Security/hayabusa/tree/imporve-speed-by-using-starts-ends-with) 
```
PS C:\tmp\hayabusa-2.1.0-win-64-bit> .\hayabusa.exe csv-timeline -d ..\hayabusa-big-evtx\ -o 1.csv --debug
Results Summary:

Events with hits / Total events: 1,593,715 / 4,817,181 (Data reduction: 3,223,466 events (66.92%))

Total | Unique detections: 1,627,284 | 150
Total | Unique critical detections: 0 (0.00%) | 0 (0.00%)
Total | Unique high detections: 12,044 (0.74%) | 20 (13.33%)
Total | Unique medium detections: 11,118 (0.68%) | 38 (25.33%)
Total | Unique low detections: 1,053,623 (64.75%) | 42 (28.00%)
Total | Unique informational detections: 550,499 (33.83%) | 50 (33.33%)

...

Saved file: 1.csv (575.1 MB)
Elapsed time: 00:10:44.026
Rule Parse Processing Time: 00:00:20.191
Analysis Processing Time: 00:10:02.174
Output Processing Time: 00:00:21.658

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    4.7 GiB    4.7 GiB   83.0 MiB    4.6 GiB
 committed:    4.3 GiB   57.6 GiB   53.4 GiB    4.1 GiB
```

[This PR](https://github.com/Yamato-Security/hayabusa/tree/improve-speed-by-reduce-regex-match-when-pipe)
```
PS C:\tmp\hayabusa-2.1.0-win-64-bit> .\hayabusa.exe csv-timeline -d ..\hayabusa-big-evtx\ -o 1.csv --debug
Results Summary:

Events with hits / Total events: 1,593,715 / 4,817,181 (Data reduction: 3,223,466 events (66.92%))

Total | Unique detections: 1,627,284 | 150
Total | Unique critical detections: 0 (0.00%) | 0 (0.00%)
Total | Unique high detections: 12,044 (0.74%) | 20 (13.33%)
Total | Unique medium detections: 11,118 (0.68%) | 38 (25.33%)
Total | Unique low detections: 1,053,623 (64.75%) | 42 (28.00%)
Total | Unique informational detections: 550,499 (33.83%) | 50 (33.33%)
...

Saved file: 1.csv (575.1 MB)
Elapsed time: 00:10:42.195
Rule Parse Processing Time: 00:00:20.234
Analysis Processing Time: 00:10:00.446
Output Processing Time: 00:00:21.513

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    4.4 GiB    4.4 GiB   83.0 MiB    4.4 GiB
 committed:    4.0 GiB   65.0 GiB   61.2 GiB    3.8 GiB
```

### Benchmark2
I ran a benchmark using [hayabusa-sample-evtx](https://github.com/Yamato-Security/hayabusa-sample-evtx) and the results were as follows.

|Version| Elapsed time | Memory(peak)  | Events with hits / Total events | Output file size(bytes)|
|-------| ------------ | --------|--------|--------|
| [before](https://github.com/Yamato-Security/hayabusa/tree/imporve-speed-by-using-starts-ends-with)   | 00:00:16.158|  1.9 GiB | 19,606 / 47,458 | 16447790 | 
| [This PR](https://github.com/Yamato-Security/hayabusa/tree/improve-speed-by-reduce-regex-match-when-pipe)   | 00:00:15.422| 1.6 GiB| 19,606 / 47,458 | 16447790 | 

I would appreciate it if you could review🙏

### CODE REVIEW

File: src/detections/rule/matchers.rs [hunk 1]
```diff
@@ -180,13 +180,19 @@ impl LeafMatcher for AllowlistFileMatcher {
     }
 }
 
+// 正規表現マッチは遅いため、できるだけ高速なstd::stringのlen/stars_with/ends_withでマッチ判定するためのenum
```
Comment (hitenkoku) [inline src/detections/rule/matchers.rs:183]: stars_with -> starts_with

File: src/detections/rule/matchers.rs [hunk 2]
```diff
@@ -225,6 +229,56 @@ impl DefaultMatcher {
...
+        }
+        // マルチバイト文字を含む場合は、index out of boundsになるため、asciiのみ
+        if event_value_str.is_ascii() {
+            let match_result = match_str.eq_ignore_ascii_case(&event_value_str[len2 - len1..]);
+            return Some(match_result);
+        }
+        None
+    }
+
+    // ワイルドカードマッチを高速なstd::stringのlen/stars_with/ends_withに変換するための関数
```
Comment (hitenkoku) [inline src/detections/rule/matchers.rs:267]: stars_with -> starts_with

**Review by hitenkoku — APPROVED** [review]
@fukusuket  Thanks for your pull request.

LGTM

**Comment by fukusuket** [comment]
@hitenkoku 
Thank you so much for quick review🙇 I fixed typo :) https://github.com/Yamato-Security/hayabusa/pull/894/commits/30d78d692a9626720b9a1c2918fae4a9ce5dacd6

**Comment by YamatoSecurity** [comment]
@fukusuket PRありがとうございます！
Macのベンチマークではメモリが300MB減っていますが、処理時間が少し増えています。
21:54.523 -> 22:26.437
１回目はメモリ使用 7.5GB -> 7.2GB、２回目は 7.5GB -> 7.4GBで、処理時間は二回目も同じ22:26でした。
明日Windowsの方でも試してみるので、少々お待ち下さい。

**Comment by fukusuket** [comment]
@YamatoSecurity 
ベンチマーク早速取得いただきありがとうございます🙇 
マッチ速度に影響しない想定の変更でしたが、すみません！どこかでデグレさせてしまっていそうです、確認いたします！
(またデータにより遅くなるパターンがありそうでしたら、こちらのPR一旦取り下げます！)

**Comment by fukusuket** [comment]
6.1GB evtxに対して手元のMac環境で試したところ(それぞれ測定前に再起動)、以下の結果でした🤔
なので、データのパターンによりスピード劣化してしまうケースがあるということかもしれません.... 😭　

#### Environment 
- OS: macOS montery version 13.1
- Hard: Macbook Air(M1, 2020) , Memory 8GB, Core 8

[before](https://github.com/Yamato-Security/hayabusa/commit/6cb871b9928e41df9c68266f96d8b70fb29214d9) 
```
Elapsed time: 00:06:22.257
Rule Parse Processing Time: 00:00:01.861
Analysis Processing Time: 00:06:15.218
Output Processing Time: 00:00:05.178

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    4.9 GiB    4.9 GiB   83.0 MiB    4.8 GiB
 committed:    4.5 GiB   49.8 GiB   45.5 GiB    4.3 GiB
```

 [This PR](https://github.com/Yamato-Security/hayabusa/tree/improve-speed-by-reduce-regex-match-when-pipe) 
```
Elapsed time: 00:06:12.900
Rule Parse Processing Time: 00:00:01.310
Analysis Processing Time: 00:06:06.634
Output Processing Time: 00:00:04.955

Memory usage stats:
heap stats:    peak      total      freed    current       unit      count
  reserved:    4.6 GiB    4.6 GiB   83.0 MiB    4.5 GiB
 committed:    4.1 GiB   49.6 GiB   45.7 GiB    3.9 GiB
```

**Comment by YamatoSecurity** [comment]
@fukusuket Hummm, yea, it might just be my environment so I will check tonight on a different computer.

File: src/detections/rule/matchers.rs (hunk 1 above)
Comment (hitenkoku) [reply src/detections/rule/matchers.rs]: resolved by 30d78d6

File: src/detections/rule/matchers.rs (hunk 2 above)
Comment (hitenkoku) [reply src/detections/rule/matchers.rs]: resolved by 30d78d6

**Review by YamatoSecurity — APPROVED** [review]
LGTM!
When checking on my Windows environment.
Speed improvement: 25:25 -> 24:45
Memory reduction: 7.6GB -> 7.2GB

**Comment by fukusuket** [comment]
Thank you for your review and benchmark :)

**Pull request merged**